### ISSUE TITLE
evaluator: structural cycle not terminated by null disjunction

### ISSUE THREAD

**Issue by disintegrator** [cue-lang/cue#1827]
### What version of CUE are you using (`cue version`)?

<pre>
$ cue version
cue version v0.4.3 darwin/arm64
</pre>

### Does this issue reproduce with the latest release?

Yes

#### Background

We recently added support for emit a CUE schema from Benthos that users can use to write safer stream processing configs. These look like this:

```cue
package benthos

#Config & {
  input: {
    generate: {
      mapping: """
      root = { "message": "Hello, CUE!" }
      """
    }
  }


  pipeline: {
    processors: [
      {
        bloblang: """
        root = this
        root.id = uuid_v4()
        """
      }
    ]
  }

  output: {
    gcp_pubsub: {
      project: "gcp-sample-project"
      topic: "flaky-topic"
    }
  }
}
```

Briefly, this contrived pipeline would generate messages on an interval that get enriched with a UUID and then published to a GCP PubSub topic.

Benthos also has higher-order components (in a manner of speaking). So, for example, if that gcp_pubsub output is failure-prone we can wrap it with mechanism for max retries:

```cue
package benthos

#Config & {
  // ... omitted for brevity ...

  output: {
    retry: {
      max_retries: 3
      output: gcp_pubsub: {
        project: "gcp-sample-project"
        topic: "flaky-topic"
      }
    }
  }
}
```

#### Issue

We have modelled all components in Benthos as CUE values that you can obtain by running `benthos list --format cue` and storing the output of that command as part of your CUE project.

As of Benthos v4.4.1, the CUE schema that is getting generated is invalid out of the box due to a structural cycle which can be detected by running `cue vet`.

What is puzzling me is that we've tried to break this cycle in the CUE generation code by forming a disjunction with `null` and yet it's not working as intended...

I have a fix ready to merge which gets rid of the null disjunction and instead uses optional fields which works https://github.com/benthosdev/benthos/pull/1355

I've attached both generated outputs to this issue to make it easier to debug (extension needed to be `.txt` so I can upload to github):

[benthos-pr1355.cue.txt](https://github.com/cue-lang/cue/files/9232652/benthos-pr1355.cue.txt)
[benthos-v4.cue.txt](https://github.com/cue-lang/cue/files/9232653/benthos-v4.cue.txt)

### What did you expect to see?

The null disjunction in `benthos-v4.cue.txt:1909` should have been sufficient to break the structural cycle.

### What did you see instead?

`cue vet` reports a structural cycle:

```
44.retry: structural cycle:
    ./benthos-v4.cue:2066:13
```

**Comment by myitcv** [comment]
Thanks for raising, @disintegrator!

Noting per our discussion yesterday that we are looking into this. In particular, I'm digging in to see whether this isn't a duplicate of https://github.com/cue-lang/cue/issues/1608 (note my most recent comment following our Slack exchange).

Just noting in passing that pattern of using `#AllOutputs` to create a disjunction in the form of `#Output` will likely be better achieved using a one-of. That will be possible once we land the revised version of https://github.com/cue-lang/cue/issues/822 (coming real soon now!) and #943. See `numexist()` as proposed in #943.

**Comment by myitcv** [comment]
Analysis thus far: #1608 appears to be partially related, but it's not the whole picture. https://review.gerrithub.io/c/cue-lang/cue/+/542674 fixes #1608, but does not fix `benthos-v4.cue.txt`. There appear to be two further issues going on related to that repro: #1881 and #1882. So we'll look into those first, then likely return to https://review.gerrithub.io/c/cue-lang/cue/+/542674 and #1608, then come back here to check again!

**Comment by disintegrator** [comment]
Thanks for the updates @myitcv! I'll track the other issues as well in the mean time.

**Comment by mvdan** [comment]
All the issues and CLs that Paul linked above have been fixed and merged at this point, including https://github.com/cue-lang/cue/issues/1882 which we appear to have fixed separately at some point recently.

@disintegrator could you please see if your bug still reproduces with the latest CUE master version? For example, I am on 48207fb5fbfd85784c0560ea32c559183b49f475.

**Comment by disintegrator** [comment]
@mvdan, @myitcv et al. sorry for the late reply. I'm testing this out now against 0.5.0 and noticing that `cue vet` and `cue export` now hangs indefinitely. I think I've seen it return once after a long wait. I've created a small reproducer here:

https://github.com/disintegrator/cue-1827

It was working with 0.4.3 after we merged my Benthos fix, mentioned in issue description, a while back. So I think the issue is possibly within CUE.

**Comment by mvdan** [comment]
Thanks for the update. Could you please clarify in the README exactly what commands you ran?

I'll take a look when I have some time and report back. Worth noting that we are aware of a few performance issues at the moment, which get particularly worse if you use a combination of disjunctions, optional fields, or deeply nested references. v0.7 will mainly be about that sort of performance work: https://github.com/cue-lang/cue/discussions/2335

**Comment by disintegrator** [comment]
@mvdan I've added a README.md with the specific command and environments tested. Also included a run.sh which runs on `good.cue` then `bad.cue`.

**Comment by disintegrator** [comment]
@mvdan I had some manual edits to benthos.cue that I've removed in https://github.com/disintegrator/cue-1827/commit/7150b6501aac826cda4a45b17d5f53b99d58df36.

The issue still persists but the schema is faithful to what comes out of `benthos list --format cue`.

**Comment by disintegrator** [comment]
@mvdan sorry for the flurry of messages... I kept investigating this issue and identified the bug in the Benthos code that was causing it. I've submitted a fix to Benthos (https://github.com/benthosdev/benthos/pull/1894) and I've demonstrated the effect of the fix here if you'd like to take a look: https://github.com/disintegrator/cue-1827/pull/1

I think it's safe to close this issue. Perhaps we can have another issue where CUE can warn/fail if it detects a problematic, recursive definition.

**Comment by mvdan** [comment]
Not at all, thank you for digging into this! I agree that opening a new issue would be best, since that appears to be quite different to the original one in this thread. I'll let you close this one out when you've created the new one, so that we don't forget about it :)

**Comment by disintegrator** [comment]
Submitted a feature request off the back of this issue: https://github.com/cue-lang/cue/issues/2398
